### PR TITLE
Document object properties

### PR DESCRIPTION
**Pull request by Minetest-j45**
This is still WIP and there are quite a few things I wasn't sure about, would be nice if someone else could help me with those things as I have already been looking through the engine code trying to figure them out for a while.

### CODE REVIEW

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:11]: What if u16 is exceeded? Does it overflow (wrap around) or is it clamped?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:9]: Please use monospace for all properties

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:19]: f32?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:37]: from the object center/pos

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:49]: Enumerate the visuals please

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:57]: Wanted value is `vector` (`{x = x, y = y, z = z}`). Important distinction.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:69]: Is it really a list and not `{x = x, y = y}`?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:73]: Same here

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:85]: f32 again?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:89]: f32 again I'd assume

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:101]: From 0 to minetest.LIGHT_MAX IIRC?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:109]: ColorSpec?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:113]: ColorSpec?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:129]: ItemString?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:145]: Will be appended to all current textures for the duration of the damage flash.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:121]: Text displayed when object is pointed at.

File: doc/object_properties.adoc
Comment (Minetest-j45) [reply doc/object_properties.adoc:11]: im not sure what the code actually does about this as I'm not very experienced in C++, you can see the relevant line: https://github.com/minetest/minetest/blob/master/src/script/common/c_content.cpp#L198

Comment (appgurueu) [reply doc/object_properties.adoc:11]:
```cpp
#define rangelim(d, min, max) ((d) < (min) ? (min) : ((d) > (max) ? (max) : (d)))
```
Looks like it's clamped

Comment (Minetest-j45) [reply doc/object_properties.adoc:11]: should i add a little bit to the end of the sentence mentioning that?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:9]: Add them please

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:17]: And what about the other properties? Most of them are clamped as well. Not only for max, but also for min.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:30]: And what's the type? `f32`?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:86]: Just `vector`

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:91]: String list.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:45]: "Whether the object collides with other objects." would be more consistent.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:210]: Remove the second part of the sentence. Order of the hash part is never important in Lua - this belongs somewhere else, in a Lua primer maybe. Consider noting that extraneous entries will be ignored, but might be interpreted incorrectly by future Minetest versions (which is why they should be prefixed with an underscore). Also please note that this uses the regular indexing, so metatables work as expected.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:102]: My intention was that field type would be the Lua value, and internal type would be the C++ value if appropriate to expose to the documentation.

**Review by appgurueu — CHANGES_REQUESTED** [review]
Mostly fine.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:134]: How is this different than `stepheight`?

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:149]: Glow needs a description.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:163]: Needs a description.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:9]: Should we add in default values now before merging?

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:40]: "Whether the object collides with nodes that are `walkable`" would be more consistent.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:46]: You need a newline before this, or it interprets it as part of the paragraph, so it doesn't display correctly.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:147]: Describing what backface culling is might be good for those who aren't experienced in graphics terms.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:161]: Need newline before TIP.

File: doc/object_properties.adoc
Comment (JosiahWI) [inline doc/object_properties.adoc:172]: This is very confusing to me.

**Review by JosiahWI — CHANGES_REQUESTED** [review]

File: doc/object_properties.adoc
Comment (Minetest-j45) [reply doc/object_properties.adoc:9]: probably

Comment (Minetest-j45) [reply doc/object_properties.adoc:9]: where would i find them?

Comment (JosiahWI) [reply doc/object_properties.adoc:11]: object_properties.h is a good place to start, but be careful - the internal defaults there may be overridden somewhere else.

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:61]: A note about it being the same format as a nodebox cuboid could be useful.

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:92]: These are separate thoughts.

```suggestion
Visual size multipliers. Wanted value is `{x = x, y = y, z = z}`. If `z` is not provided, `x` will be used for the `z` value.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:132]: What does "step up" mean if I'm not familiar with engine terminology?

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:153]: What bit length?

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:167]: The string"false" or the `bool` value `false`? If the latter, it should be "`ColorSpec` or `bool`.

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:171]: This is inconsistent with the one above it.

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:175]: Same question as the property above it.

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:178]: What bit length?

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:190]:
```suggestion
Whether the object is saved statically, if not, it will be deleted when the world block is unloaded.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:205]: This is slightly clearer IMO, but not by much.

```suggestion
Controls whether or not diffuse lighting is applied to the object.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:210]:
```suggestion
Controls whether or not the object is visible on the minimap.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:225]: Use a named example block please.

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:29]:
```suggestion
Sets the camera height. Measured in nodes above the player's feet.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:42]:
```suggestion
Controls whether the object collides with nodes that are `walkable`.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:35]:
```suggestion
Sets the field of view of the player when they zoom. Setting to `0` disables zooming.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:47]:
```suggestion
Controls whether the object collides with other objects.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:66]:
```suggestion
Controls whether the object can be pointed at.
```

File: doc/object_properties.adoc
Comment (benrob0329) [inline doc/object_properties.adoc:80]:
```suggestion
Sets what the object will look like.
```

**Review by benrob0329 — CHANGES_REQUESTED** [review]
Incomplete review but there's a lot here.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:49]: Perhaps also worth noting that this will work for players as well, but exhibits strange bugs (such as this one: https://github.com/minetest/minetest/issues/11783)

Comment (Minetest-j45) [reply doc/object_properties.adoc:49]: is this not implied by the `Properties Applying to All Objects` subheading?

Comment (appgurueu) [reply doc/object_properties.adoc:49]: > is this not implied by the `Properties Applying to All Objects` subheading?

The former yes, the latter no.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:170]: Unnecessary, `false` is always a bool.

Comment (Minetest-j45) [reply doc/object_properties.adoc:170]: https://github.com/minetest/minetest_docs/pull/12#discussion_r778418292 ...

Comment (appgurueu) [reply doc/object_properties.adoc:170]: > [#12 (comment)](https://github.com/minetest/minetest_docs/pull/12#discussion_r778418292) ...

Strings should IMO be quoted.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:15]: Are you sure about this?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:22]: Again, sure?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:43]: Lua API has a different default: https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L7168

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:55]: experience*

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:60]: Is the Lua API wrong? https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L7174-L7175

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:66]: https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L7174-L7175

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:108]: `list` was better, IMO.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:160]: 0 according to lua_api.txt

**Review by appgurueu — CHANGES_REQUESTED** [review]
Inconsistent with lua_api.txt, missing a few default values.

**Comment by JosiahWI** [comment]
I have looked over the places where the engine sets defaults, and the defaults @appgurueu has questioned are correct as checked in. Celeron55 mentioned on IRC that the player has its own defaults, could those possibly have been confused with the defaults in object_properties.h somehow?

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:120]: `{{r = 255, g = 255, b = 255, a = 255}}`

**Review by JosiahWI — CHANGES_REQUESTED** [review]

**Review by JosiahWI — COMMENTED** [review]
Thanks, this documentation looks good, and I think we are ready to merge it once we get the required approval.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:290]: Please sort this properly. This property should be next to `automatic_face_movement_dir`.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:278]: Note a sprite, so please remove the irrelevant properties.

File: doc/object_properties.adoc
Comment (appgurueu) [inline doc/object_properties.adoc:292]: Is this an example for a player? Why set the `*_max` properties if it's not?

**Review by appgurueu — CHANGES_REQUESTED** [review]
Please fix the following:

* Incorrectly specified default values
* Some overly vague statements
* `nametag` behavior is described incorrectly
* `glow` is documented insufficiently
* Self-contradicting example including usage of ignored properties
* A handful of nitpicks (mostly formatting-related)

File: doc/object_properties.adoc
Comment (Minetest-j45) [reply doc/object_properties.adoc:259]: the leading dot is for the named example block, I will fix the inconsistent title case

**Review by appgurueu — APPROVED** [review]
Looks good enough to me. Further improvements can be made in a follow-up PR if necessary. Please squash this when committing.

**Pull request merged**
